lib/opal/rewriters/logical_operator_assignment.rb

```diff
@@ -0,0 +1,199 @@
+require 'opal/rewriters/base'
+
+module Opal
+  module Rewriters
+    class LogicalOperatorAssignment < Base
+      class AbstractHandler < self
+        def self.get(lhs, rhs)
+          raise NotImplementedError
+        end
+
+        def self.set(lhs, rhs)
+          raise NotImplementedError
+        end
+
+        def self.process(lhs, rhs, root_type)
+          s(root_type,
+            get(lhs, rhs),
+            set(lhs, rhs))
+        end
+
+        def self.new_temp
+          @@counter ||= 0
+          @@counter += 1
+          :"$or_equals_lhs_recvr_tmp_#{@@counter}"
+        end
+
+        def self.reset_tmp_counter!
+          @@counter = 0
+        end
+      end
+
+      # Takes    `lhs ||= rhs`
+      # Produces `lhs || lhs = rhs`
+      class LocalVariableHandler < AbstractHandler
+        def self.get(lhs, rhs)
+          lhs.updated(:lvar) # lhs
+        end
+
+        def self.set(lhs, rhs)
+          lhs.updated(:lvasgn, [*lhs, rhs]) # lhs = rhs
+        end
+      end
```

Comment:
Let's forget the registry for now, I think I still like the more concise first version that uses `ivars`:

```rb
class GetSetHandler < AbstractHandler
  def self.get(lhs, rhs)
    lhs.updated(@get_type) # lhs
  end
  def self.set(lhs, rhs)
    lhs.updated(@set_type, [*lhs, rhs]) # lhs = rhs
  end
end

# Takes    `lhs ||= rhs`
# Produces `lhs || lhs = rhs`
class LocalVariableHandler < GetSetHandler
  @get_type = :lvar
  @set_type = :lvasgn
end
```